File: Leetcode/1080.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sufficientSubset(root *TreeNode, limit int) *TreeNode {
    if root.Left==nil && root.Right==nil{ //если узел - лист, проверяем условие
        if root.Val<limit{
            return nil
        }else{
            return root
        }
    }
    if root.Left!=nil{ // если есть левый потомок, то запускаем рекурсию с уменьшением limit'а
        root.Left=sufficientSubset(root.Left, limit-root.Val)
    }
    if root.Right!=nil{ // если есть правый потомок, то запускаем рекурсию с уменьшением limit'а
        root.Right=sufficientSubset(root.Right,limit-root.Val)
    }
    if root.Left!=nil || root.Right!=nil{ //если после рекурсий у узла не осталось потомков - этот узел insufficient
        return root
    }
    return nil
}

File: 1080.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} limit
 * @return {TreeNode}
 */
var sufficientSubset = function (root, limit) {
  DFS = (root, sum) => {
    if (!root) return null;
    sum += root.val;
    if (!root.left && !root.right) {
      if (sum < limit) return null;
      return root;
    }
    root.left = DFS(root.left, sum);
    root.right = DFS(root.right, sum);
    if (!root.left && !root.right) {
      return null;
    }
    return root;
  };
  return DFS(root, 0);
};
